sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/ProxyAuthenticationHandler.java

```diff
@@ -0,0 +1,313 @@
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.proxy.ProxyConnectException;
+import io.netty.handler.proxy.ProxyHandler;
+import io.netty.util.AttributeKey;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static com.azure.core.util.AuthorizationChallengeHandler.PROXY_AUTHENTICATE;
+import static com.azure.core.util.AuthorizationChallengeHandler.PROXY_AUTHENTICATION_INFO;
+import static com.azure.core.util.AuthorizationChallengeHandler.PROXY_AUTHORIZATION;
+
+/**
+ * This class handles authorizing requests being sent through a proxy which require authentication.
+ */
+public final class ProxyAuthenticationHandler extends ProxyHandler {
+    private static final String VALIDATION_ERROR_TEMPLATE = "The '%s' returned in the 'Proxy-Authentication-Info' "
+        + "header doesn't match the value sent in the 'Proxy-Authorization' header. Sent: %s, received: %s.";
+
+    private static final AttributeKey<String> PROXY_AUTHORIZATION_KEY = AttributeKey.newInstance("ProxyAuthorization");
+
+    private static final String NONE = "none";
+    private static final String HTTP = "http";
+
+    private static final String CNONCE = "cnonce";
+    private static final String NC = "nc";
+
+    private static final String AUTH_BASIC = "basic";
+    private static final String AUTH_DIGEST = "digest";
+
+    /*
+     * Proxies use 'CONNECT' as the HTTP method.
+     */
+    private static final String PROXY_METHOD = HttpMethod.CONNECT.name();
+
+    /*
+     * Proxies are always the root path.
+     */
+    private static final String PROXY_URI_PATH = "/";
+
+    /*
+     * Digest authentication to a proxy uses the 'CONNECT' method, these can't have a request body.
+     */
+    private static final Supplier<byte[]> PROXY_ENTITY_BODY = () -> new byte[0];
+
+    private final ClientLogger logger = new ClientLogger(ProxyAuthenticationHandler.class);
+
+    private final AuthorizationChallengeHandler challengeHandler;
+    private final AtomicReference<ChallengeHolder> proxyChallengeHolderReference;
+    private final HttpClientCodec codec;
+    private final AtomicReference<String> authScheme;
+
+    private HttpResponseStatus status;
+
+    public ProxyAuthenticationHandler(InetSocketAddress proxyAddress, AuthorizationChallengeHandler challengeHandler,
+        AtomicReference<ChallengeHolder> proxyChallengeHolderReference) {
+        super(proxyAddress);
+
+        this.challengeHandler = challengeHandler;
+        this.proxyChallengeHolderReference = proxyChallengeHolderReference;
+
+        this.codec = new HttpClientCodec();
+        this.authScheme = new AtomicReference<>();
```

Comment:
When is authScheme updated/set? It isn't clear to me where this happens or why we need an atomic reference for it.

Reply:
Looks like I forgot to add this functionality, it is used in logging for Netty. It doesn't need to be an `AtomicReference` as it will only exist in the `ProxyAuthenticationHandler` which isn't shared across Netty pipelines, which should mean that it won't be accessed by multiple threads at once.

Reply:
Added functionality to set `authScheme`